src/freenet/clients/fcp/FCPPluginClientTracker.java

```diff
@@ -0,0 +1,212 @@
+     * @throws IOException
+     *             If there has been no client with the given ID or if it has disconnected already.
+     */
+    public FCPPluginClient getClient(UUID clientID) throws IOException {
+        FCPPluginClientWeakReference ref = null;
+        
+        clientsByIDLock.readLock().lock();
+        try {
+            ref = clientsByID.get(clientID);
+        } finally {
+            clientsByIDLock.readLock().unlock();
+        }
+        
+        FCPPluginClient client = ref != null ? ref.get() : null;
+        
+        if(client == null) {
+            throw new IOException("FCPPluginClient not found, maybe it has disconnected already."
+                + " ClientID: " + clientID);
+        }
+        
+        return client;
+    }
+
+
+    /**
+     * You must call {@link #start()} afterwards!
+     */
+    public FCPPluginClientTracker() {
+        super("FCPPluginClientTracker Garbage-collector",
+            NativeThread.PriorityLevel.MIN_PRIORITY.value, true);
+        setDaemon(true);
+    }
+
+    /**
+     * Garbage-collection thread: Pools {@link #disconnectedClientsQueue} for clients whose
+     * {@link WeakReference} has been nulled and removes them from the {@link #clientsByID}
+     * {@link TreeMap}.
+     */
+    @Override
+    public void realRun() {
+        while(true) {
+            try {
+                FCPPluginClientWeakReference disconnectedClient
+                = (FCPPluginClientWeakReference)disconnectedClientsQueue.remove();
+
+                clientsByIDLock.writeLock().lock();
+                try {
+                    FCPPluginClientWeakReference removedFromTree
+                    = clientsByID.remove(disconnectedClient.clientID);
+
+                    assert(disconnectedClient == removedFromTree);
+                    if(logMINOR) {
+                        Logger.minor(this, "Garbage-collecting disconnected client: " +
+                            "remaining clients = " + clientsByID.size() +
+                            "; client ID = " + disconnectedClient.clientID);
+                    }
+                } finally {
+                    clientsByIDLock.writeLock().unlock();
+                }
+            } catch(InterruptedException e) {
+                // We did setDaemon(true), which causes the JVM to exit even if the thread is still
+                // running: Daemon threads are force terminated during shutdown.
+                // Thus, this thread does not need an exit mechanism, it can be an infinite loop. So
+                // nothing should try to terminate it by InterruptedException. If it does happen
+                // nevertheless, we honor it by exiting the thread, because interrupt requests
+                // should never be ignored, but log it as an error.
+                Logger.error(this,
+                    "Thread interruption requested even though this is a daemon thread!", e);
+                throw new RuntimeException(e);
```

Comment:
I recently told you that ThreadGroup can be used to add a handler for uncaught exceptions in the case of standard Java class Thread. I would suspect that this is also possible with our NativeThread, and it is possible that something will listen.  Thus I would like to keep it that way. It is an error after all: Nobody is supposed to shutdown the garbage-collector, this will cause memory leaks. So we do what you do upon error - throw.
ACK to keep as is?